File: go/bounded.go
package consistenthash

import (
	"fmt"
	"math"
	"sort"
)

// BoundedLoad is consistent hashing with bounded loads (Google, 2017): keys are
// placed clockwise on the ring, but no node may exceed ceil((1+Epsilon)*keys/nodes).
// When the clockwise owner is full, the key overflows to the next node under cap, so
// no node exceeds its fair share by more than Epsilon. Placement depends on load, so
// it is stateful and order-sensitive.
type BoundedLoad struct {
	replicas int
	epsilon  float64
	nodes    []string
	ring     map[uint32]string
	sorted   []uint32
	load     map[string]int
	assigned map[string]string
}

// NewBoundedLoad builds a bounded-load placer over nodes with the given replica count
// and epsilon (the allowed overshoot above the mean, e.g. 0.25 for +25%).
func NewBoundedLoad(nodes []string, replicas int, epsilon float64) *BoundedLoad {
	if epsilon < 0 {
		panic("epsilon must be non-negative")
	}
	b := &BoundedLoad{
		replicas: replicas,
		epsilon:  epsilon,
		nodes:    dedupeSorted(nodes),
		ring:     make(map[uint32]string),
		load:     make(map[string]int),
		assigned: make(map[string]string),
	}
	b.buildRing()
	return b
}

// Get returns the node owning key, respecting the per-node cap. It is idempotent for
// a key already assigned.
func (b *BoundedLoad) Get(key string) string {
	if len(b.sorted) == 0 {
		return ""
	}
	if node, ok := b.assigned[key]; ok {
		return node
	}
	capacity := b.capacity()
	start := b.firstClockwise(Fnv1a32(key))
	n := len(b.sorted)
	for i := 0; i < n; i++ {
		node := b.ring[b.sorted[(start+i)%n]]
		if b.load[node] < capacity {
			b.assigned[key] = node
			b.load[node]++
			return node
		}
	}
	// Every node at cap (only with a tiny epsilon): fall back to first clockwise.
	node := b.ring[b.sorted[start%n]]
	b.assigned[key] = node
	b.load[node]++
	return node
}

// Loads returns a copy of the current per-node key counts.
func (b *BoundedLoad) Loads() map[string]int {
	out := make(map[string]int, len(b.load))
	for k, v := range b.load {
		out[k] = v
	}
	return out
}

// Nodes returns the nodes in sorted order.
func (b *BoundedLoad) Nodes() []string {
	out := make([]string, len(b.nodes))
	copy(out, b.nodes)
	return out
}

// Len is the number of nodes.
func (b *BoundedLoad) Len() int { return len(b.nodes) }

func (b *BoundedLoad) capacity() int {
	total := len(b.assigned) + 1
	return int(math.Ceil((1 + b.epsilon) * float64(total) / float64(len(b.nodes))))
}

func (b *BoundedLoad) buildRing() {
	for _, node := range b.nodes {
		for i := 0; i < b.replicas; i++ {
			slot := Fnv1a32(fmt.Sprintf("%s#%d", node, i))
			for {
				if _, taken := b.ring[slot]; !taken {
					break
				}
				slot++
			}
			b.ring[slot] = node
		}
	}
	b.sorted = b.sorted[:0]
	for slot := range b.ring {
		b.sorted = append(b.sorted, slot)
	}
	sort.Slice(b.sorted, func(i, j int) bool { return b.sorted[i] < b.sorted[j] })
}

func (b *BoundedLoad) firstClockwise(h uint32) int {
	idx := sort.Search(len(b.sorted), func(i int) bool { return b.sorted[i] > h })
	if idx == len(b.sorted) {
		return 0
	}
	return idx
}

File: rust/src/bounded.rs
use crate::fnv1a_32;
use std::collections::HashMap;

/// Consistent hashing with bounded loads (Google, 2017): keys are placed clockwise on
/// the ring, but no node may exceed `ceil((1+epsilon)*keys/nodes)`. A full node
/// overflows to the next under cap, so no node exceeds its fair share by more than
/// `epsilon`. Placement depends on load, so it is stateful and order-sensitive.
pub struct BoundedLoad {
    epsilon: f64,
    nodes: Vec<String>,
    ring: HashMap<u32, String>,
    sorted: Vec<u32>,
    load: HashMap<String, usize>,
    assigned: HashMap<String, String>,
}

impl BoundedLoad {
    pub fn new(nodes: &[&str], replicas: usize, epsilon: f64) -> Self {
        assert!(epsilon >= 0.0, "epsilon must be non-negative");
        let mut sorted: Vec<String> = nodes.iter().map(|n| n.to_string()).collect();
        sorted.sort();
        sorted.dedup();
        let mut b = BoundedLoad {
            epsilon,
            nodes: sorted,
            ring: HashMap::new(),
            sorted: Vec::new(),
            load: HashMap::new(),
            assigned: HashMap::new(),
        };
        b.build_ring(replicas);
        b
    }

    pub fn get(&mut self, key: &str) -> Option<String> {
        if self.sorted.is_empty() {
            return None;
        }
        if let Some(node) = self.assigned.get(key) {
            return Some(node.clone());
        }
        let cap = self.capacity();
        let start = self.first_clockwise(fnv1a_32(key));
        let n = self.sorted.len();
        for i in 0..n {
            let node = self.ring[&self.sorted[(start + i) % n]].clone();
            if *self.load.get(&node).unwrap_or(&0) < cap {
                self.assign(key, &node);
                return Some(node);
            }
        }
        let node = self.ring[&self.sorted[start % n]].clone();
        self.assign(key, &node);
        Some(node)
    }

    pub fn loads(&self) -> HashMap<String, usize> {
        self.load.clone()
    }

    pub fn nodes(&self) -> Vec<String> {
        self.nodes.clone()
    }

    pub fn len(&self) -> usize {
        self.nodes.len()
    }

    pub fn is_empty(&self) -> bool {
        self.nodes.is_empty()
    }

    fn assign(&mut self, key: &str, node: &str) {
        self.assigned.insert(key.to_string(), node.to_string());
        *self.load.entry(node.to_string()).or_insert(0) += 1;
    }

    fn capacity(&self) -> usize {
        let total = self.assigned.len() + 1;
        ((1.0 + self.epsilon) * total as f64 / self.nodes.len() as f64).ceil() as usize
    }

    fn build_ring(&mut self, replicas: usize) {
        for node in &self.nodes {
            for i in 0..replicas {
                let mut slot = fnv1a_32(&format!("{node}#{i}"));
                while self.ring.contains_key(&slot) {
                    slot = slot.wrapping_add(1);
                }
                self.ring.insert(slot, node.clone());
            }
        }
        self.sorted = self.ring.keys().copied().collect();
        self.sorted.sort_unstable();
    }

    fn first_clockwise(&self, h: u32) -> usize {
        match self.sorted.partition_point(|&s| s <= h) {
            i if i == self.sorted.len() => 0,
            i => i,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn caps_hot_nodes() {
        let ns: Vec<String> = (0..5).map(|i| format!("node-{i}")).collect();
        let refs: Vec<&str> = ns.iter().map(|s| s.as_str()).collect();
        let mut b = BoundedLoad::new(&refs, 200, 0.25);
        for i in 0..5000 {
            b.get(&format!("key-{i}"));
        }
        let loads = b.loads();
        let total: usize = loads.values().sum();
        let peak = *loads.values().max().unwrap();
        let mean = total as f64 / ns.len() as f64;
        assert!(peak as f64 <= 1.25 * mean + 1.0, "peak {peak} exceeded cap");
    }

    #[test]
    fn idempotent_per_key() {
        let mut b = BoundedLoad::new(&["a", "b", "c"], 100, 0.5);
        let first = b.get("user-42");
        assert_eq!(first, b.get("user-42"));
    }
}
